ref-farming/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+/*!
+* Ref-Farming
+*
+* lib.rs is the main entry point.
+*/
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::{ValidAccountId};
+use near_sdk::collections::{LookupMap, UnorderedMap};
+use near_sdk::{env, near_bindgen, Balance, AccountId, PanicOnDefault};
+
+use crate::farm::{Farm, FarmId};
+use crate::farm_seed::{FarmSeed, SeedId};
+use crate::farmer::Farmer;
+
+// for simulator test
+pub use crate::simple_farm::HRSimpleFarmTerms;
+pub use crate::view::FarmInfo;
+
+
+mod utils;
+mod errors;
+mod farmer;
+mod token_receiver;
+mod farm_seed;
+mod farm;
+mod simple_farm;
+mod storage_impl;
+
+mod actions_of_farm;
+mod actions_of_seed;
+mod actions_of_reward;
+mod view;
+
+/// sodu module is used to debug and testing,
+/// remove this module in release version
+mod sudo;
+
+near_sdk::setup_alloc!();
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+
+    // owner of this contract
+    owner_id: AccountId,
+    
+    // record seeds and the farms under it.
+    seeds: UnorderedMap::<SeedId, FarmSeed>,
+
+    // each farmer has a structure to describe
+    farmers: LookupMap<AccountId, Farmer>,
+
+    // for statistic
+    farmer_count: u64,
+    farm_count: u64,
+    reward_info: UnorderedMap::<AccountId, Balance>,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(owner_id: ValidAccountId) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            owner_id: owner_id.into(),
+            seeds: UnorderedMap::new(b"s".to_vec()),
```

Comment:
let's use StorageKeys model: https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/lib.rs#L33

Reply:
Wow, cool, neat code. I will follow that.

Reply:
I leave the code in comment there, cause the main project using near-rs-sdk 3.0.0, where StorageKeys is not supported. I want this PR is limited in ref-farming sub folder, so I won't change Cargo.html of main project. Once the main project upgrade to 3.1.0, I will refresh the code correspondingly.